substrate/frame/support/src/traits/try_runtime.rs

```diff
@@ -63,26 +63,47 @@ impl Default for Select {
 /// decoded bytes if `Ok(_)`.
 pub trait TryDecodeEntireStorage {
 	/// Decode the entire data under the given storage, returning `Ok(bytes_decoded)` if success.
-	fn try_decode_entire_state() -> Result<usize, &'static str>;
+	fn try_decode_entire_state() -> Result<usize, TryDecodeEntireStorageError>;
 }
 
 #[cfg_attr(all(not(feature = "tuples-96"), not(feature = "tuples-128")), impl_for_tuples(64))]
 #[cfg_attr(all(feature = "tuples-96", not(feature = "tuples-128")), impl_for_tuples(96))]
 #[cfg_attr(feature = "tuples-128", impl_for_tuples(128))]
 impl TryDecodeEntireStorage for Tuple {
-	fn try_decode_entire_state() -> Result<usize, &'static str> {
+	fn try_decode_entire_state() -> Result<usize, TryDecodeEntireStorageError> {
 		let mut len = 0usize;
 		for_tuples!( #( len = len.saturating_add(Tuple::try_decode_entire_state()?); )* );
 		Ok(len)
 	}
 }
 
+/// A value could not be decoded.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct TryDecodeEntireStorageError {
+	/// The key of the undecodable value.
+	key: Vec<u8>,
+	/// The raw value.
+	raw: Option<Vec<u8>>,
+	/// The storage info of the key.
+	info: StorageInfo,
```

Comment:
I took he liberty to add a custom error type. In the future we can maybe use a custom error instead of `DispatchError`, since it only supports `&'static str`.  
Maybe we can create a wrapper type that can be created from a `DispatchError`.